### PR TITLE
BS-151 | Bulk FHIR export of Anonymised Patient Data from Bahmni (Redact)

### PR DESCRIPTION
**Pull request by manimaarans**
## Requirements

- [x] This PR has a proper title that briefly describes the work done
- [x] I have squashed / amended the comments to make it more redable
- [x] I have included link to all the JIRA ticket('s)
- [x] I wrote the code as a pair or atleast performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## JIRA tickets

https://bahmni.atlassian.net/browse/BS-151

### CODE REVIEW

File: omod/src/main/resources/config.xml [hunk 1]
```diff
@@ -93,6 +93,10 @@
 		<name>Export Patient Data</name>
 		<description>Ability to export patient data in FHIR format</description>
 	</privilege>
+	<privilege>
+		<name>Export Non Anonymised Patient Data</name>
+		<description>Ability to export non anonymised patient data in FHIR format</description>
```
Comment (gsluthra) [inline omod/src/main/resources/config.xml:98]: Ability to bulk-export patient data that has NOT been anonymised, in FHIR JSON format.

File: omod/src/main/resources/config.xml [hunk 2]
```diff
@@ -111,4 +115,10 @@
 		<defaultValue></defaultValue>
 		<description>Properties file path for procedure template</description>
 	</globalProperty>
+
+	<globalProperty>
+		<property>fhir.export.anonymise.config.path</property>
+		<defaultValue></defaultValue>
+		<description>Properties file path for fhir export anonymisation</description>
```
Comment (gsluthra) [inline omod/src/main/resources/config.xml:122]: Path of the config file that controls the anonymisation mechanism during bulk-export of patient data.

**Review by gsluthra — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/fhirExtension/export/impl/PatientExport.java [hunk 3]
```diff
@@ -16,17 +17,25 @@ public class PatientExport implements Exporter {
...
+	public List<IBaseResource> export(String startDate, String endDate, boolean isAnonymise) {
 		DateRangeParam lastUpdated = getLastUpdated(startDate, endDate);
 		PatientSearchParams patientSearchParams = new PatientSearchParams(null, null, null, null, null, null, null, null,
 		        null, null, null, null, null, lastUpdated, null, null);
 		IBundleProvider iBundleProvider = fhirPatientService.searchForPatients(patientSearchParams);
-		return iBundleProvider.getAllResources();
+		return isAnonymise ? anonymise(iBundleProvider.getAllResources()) : iBundleProvider.getAllResources();
+	}
+	
+	private List<IBaseResource> anonymise(List<IBaseResource> iBaseResources) {
```
Comment (vijayanandtwks) [inline api/src/main/java/org/openmrs/module/fhirExtension/export/impl/PatientExport.java:37]: All implementations of Exporter have this redundant method 
List<IBaseResource> anonymise(List<IBaseResource> iBaseResources)
Add additional parameter i.e. resource type and move it to default method in the interface

File: api/src/main/java/org/openmrs/module/fhirExtension/export/anonymise/config/AnonymisedResourceConfig.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package org.openmrs.module.fhirExtension.export.anonymise.config;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class AnonymisedResourceConfig {
```
Comment (vijayanandtwks) [inline api/src/main/java/org/openmrs/module/fhirExtension/export/anonymise/config/AnonymisedResourceConfig.java:8]: Naming not consistent - anonymised, anonymiser, anonymise. Use only one style

**Comment by gsluthra** [comment]
OK from my side. thanks.

**Review by gsluthra — APPROVED** [review]

**Review by vijayanandtwks — APPROVED** [review]

**Pull request merged**